File: apps/organiclever-be/tests/integration/Steps/NatsConfigSteps.fs
module OrganicleverBe.Tests.Integration.Steps.NatsConfigSteps

open System
open TickSpec
open Xunit
open OrganicleverBe.Infrastructure.NatsClient

let mutable private previousValue: string = null
let mutable private caughtException: exn option = None

[<Given>]
let ``ORGANICLEVER_BE_NATS_URL is unset`` () =
    previousValue <- Environment.GetEnvironmentVariable("ORGANICLEVER_BE_NATS_URL")
    Environment.SetEnvironmentVariable("ORGANICLEVER_BE_NATS_URL", null)
    caughtException <- None

[<When>]
let ``organiclever-be reads its messaging configuration`` () =
    caughtException <-
        try
            requireNatsUrl () |> ignore
            None
        with ex ->
            Some ex

    Environment.SetEnvironmentVariable("ORGANICLEVER_BE_NATS_URL", previousValue)

[<Then>]
let ``startup aborts with a clear missing-variable error`` () =
    Assert.True(caughtException.IsSome)
    Assert.Contains("ORGANICLEVER_BE_NATS_URL", caughtException.Value.Message)

File: apps/organiclever-be-e2e/steps/nats-config.steps.ts
import { spawnSync, type SpawnSyncReturns } from "node:child_process";
import { expect } from "@playwright/test";
import { createBdd } from "playwright-bdd";

const { Given, When, Then } = createBdd();

let childEnvironment: NodeJS.ProcessEnv = {};
let startupResult: SpawnSyncReturns<string> | undefined;

Given("ORGANICLEVER_BE_NATS_URL is unset", async () => {
  childEnvironment = { ...process.env, ORGANICLEVER_BE_NATS_URL: "" };
  expect(childEnvironment["ORGANICLEVER_BE_NATS_URL"]).toBe("");
});

When("organiclever-be reads its messaging configuration", async () => {
  startupResult = spawnSync(
    "dotnet",
    [
      "run",
      "--project",
      "../organiclever-be/src/OrganicleverBe/OrganicleverBe.fsproj",
      "--no-build",
      "--configuration",
      "Release",
    ],
    { cwd: process.cwd(), env: childEnvironment, encoding: "utf8", timeout: 30_000 },
  );
});

Then("startup aborts with a clear missing-variable error", async () => {
  expect(startupResult).toBeDefined();
  expect(startupResult!.status).not.toBe(0);
  expect(`${startupResult!.stdout}\n${startupResult!.stderr}`).toContain("ORGANICLEVER_BE_NATS_URL");
});
